src/repository/opamHg.ml

```diff
@@ -94,13 +88,39 @@ module VCS = struct
 
   let vc_dir repo_root = OpamFilename.Op.(repo_root / ".hg")
 
-  let current_branch dir =
-    hg dir [ "branch" ] @@> function
-    | { OpamProcess.r_code = 0; OpamProcess.r_stdout = [b]; _ } -> Done (Some b)
-    | _ -> Done None
-
-  let is_dirty dir =
-    hg dir [ "stat" ] @@> fun r ->
+  let filter_opam_mark mark =
+    String.length mark < 9 || String.sub mark 0 9 <> "opam-mark"
+
+  let split_on_char sep s =
+    let r = ref [] in
+    let j = ref (String.length s) in
+    for i = String.length s - 1 downto 0 do
+      if String.unsafe_get s i = sep then begin
+        r := String.sub s (i + 1) (!j - i - 1) :: !r;
+        j := i
+      end
+    done;
+    String.sub s 0 !j :: !r
+
+  let current_branch repo_root =
+    hg repo_root [ "identify"; "--bookmarks" ] @@> fun r ->
+    OpamSystem.raise_on_process_error r;
+    match r.OpamProcess.r_stdout with
+    | [] -> Done None
+    | marks::_ ->
+        let marks = split_on_char ' ' marks in
+        let marks = List.filter filter_opam_mark marks in
+        match marks with
+        | mark::_ -> Done (Some mark)
+        | [] ->
+            hg repo_root [ "identify"; "--branch" ] @@> fun r ->
+            OpamSystem.raise_on_process_error r;
+            match r.OpamProcess.r_stdout with
+            | branch::_ when branch <> "default" -> Done (Some branch)
+            | _ -> Done None
```

Comment:
In Mercurial terminology, a "bookmark" is functionally equivalent to a Git branch. (In fact, when Git repositories are mirrored in Mercurial, remote Git branches are assigned Mercurial bookmarks.)

There is no Git concept equivalent to the "branch" in Mercurial terminology, which is an attribute of a revision, not a reference to a head. When a Mercurial branch name is used to refer to a revision, it maps to the most currently advanced head attributed with that branch name, and multiple heads are permitted in the same branch accordingly.

What I have done here is to map the fragment of the URL to Mercurial's concept of a named revision, which can be either A) a bookmark, or B) a branch name. (Actually, there are a few additional reserved names, e.g. "tip", but I don't see them as useful here.) If the bookmark exists, then the fragment maps to the bookmark. Otherwise, it maps to a branch. The usage is intuitive for a Mercurial user, as this is how named revisions are handled by the `--rev` option in the various Mercurial commands.

Reply:
Great, perfect then, and thanks for the explanations, that makes full sense now :+1: